### PR TITLE
feat: Support Transformers custom pipeline

### PR DESCRIPTION
**Pull request by ssheng**
## Description

This PR adds support for Huggingface Transformers custom pipelines as requested in https://github.com/bentoml/BentoML/issues/2534.

The Huggingface documentation instructed creation of custom pipelines by modifying the `SUPPORTED_TASKS`, see https://huggingface.co/docs/transformers/add_new_pipeline. To support saving custom pipelines, the framework code will first check the presence of the task name of the custom pipeline and save all relevant task definition information (task name, PyTorch and Tensorflow auto class names, default models, and pipeline type) to the model options. When loading pipeline, the framework code will add the custom task to the `SUPPORTED_TASKS` with the metadata saved in the model options and construct the pipeline.

## How Has This Been Tested?

Added framework test for saving and loading custom Transformers pipelines. This PR does not cover the following two tasks which I want to cover in the future.
- Refactor the framework test to use the new template.
- Refactor the framework test to use simpler mocks of PyTorch and Tensorflow models instead of using the actual models from the model hub.

Current working on adding framework tests for custom pipeline. Encountered issue with compatibility issues with Tensorflow  and Apple Silicon.

## Checklist:

- [ ] My code follows the bentoml code style, both `make format` and
  `make lint` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/_internal/frameworks/transformers.py [hunk 1]
```diff
@@ -84,13 +87,11 @@ def _check_flax_supported() -> None:  # pragma: no cover
...
-    task: str = attr.field(
-        validator=[
-            attr.validators.instance_of(str),
-            lambda instance, attribute, value: transformers.pipelines.check_task(value),  # type: ignore
-        ]
-    )
-
+    task: str = attr.field(validator=[attr.validators.instance_of(str)])
+    tf: t.Tuple(str) = (attr.field(validator=[attr.validators.instance_of(str)], default=None),)  # type: ignore
+    pt: t.Tuple(str) = (attr.field(validator=[attr.validators.instance_of(str)], default=None),)  # type: ignore
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/transformers.py:92]: I believe you shouldn't doing this. Instead you should use factory=tuple and attrs.validators.deep_iterable

File: bentoml/_internal/frameworks/transformers.py [hunk 2]
```diff
@@ -160,7 +190,35 @@ def save_model(
         name (:code:`str`):
             Name for given model instance. This should pass Python identifier check.
         pipeline (:code:`Pipeline`):
-            Instance of the Transformers pipeline to be saved.
+            Instance of the Transformers pipeline to be saved. See Transformers
+            ``module src/transformers/pipelines/__init__.py`` for more details.
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/transformers.py:194]: lets 
rovide a link to the source code here

File: bentoml/_internal/frameworks/transformers.py [hunk 3]
```diff
@@ -160,7 +190,35 @@ def save_model(
...
+        task_name (:code:`str`):
+            Name of pipeline task. If not provided, the task name will be derived from ``pipeline.task``.
+            ``task_name`` and ``task_definition`` must be both provided or both not provided.
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/transformers.py:197]: maybe another way to say this is

if task_name is provided, then task_definition is also required.

**Review by aarnphm — COMMENTED** [review]
some comments here and there

File: bentoml/_internal/frameworks/transformers.py (hunk 1 above)
Comment (ssheng) [reply bentoml/_internal/frameworks/transformers.py:92]: Good catch. I forgot to update the type after changing the type to `tuple`.

File: bentoml/_internal/frameworks/transformers.py [hunk 4]
```diff
@@ -160,7 +190,35 @@ def save_model(
...
+            ``impl`` (:code:`str`): The name of the pipeline implementation module. The name should
+            be the same as the pipeline passed in the ``pipeline`` argument.
+            ``tf`` (:code:`tuple[AnyType]`): The name of the Tensorflow auto model class. One of ``tf`` and ``pt``
+            auto model class argument is required.
+            ``pt`` (:code:`tuple[AnyType]`): The name of the PyTorch auto model class. One of ``tf`` and ``pt``
+            auto model class argument is required.
+            ``default`` (:code:`Dict[str, AnyType]`): The names of the default models, tokenizers, feature extractors, etc.
+            ``type`` (:code:`str`): The type of the pipeline, e.g. ``text``, ``audio``, ``image``, ``multimodal``.
+
+            Example:
```
Comment (aarnphm) [inline bentoml/_internal/frameworks/transformers.py:211]:
```suggestion
            .. admonition:: Example:
```

**Review by aarnphm — DISMISSED** [review]
LGTM.

**Comment by sauyon** [comment]
isn't it fine?

**Comment by aarnphm** [comment]
> isn't it fine?

frameworks tests are skipped.

**Review by aarnphm — APPROVED** [review]

**Review by aarnphm — CHANGES_REQUESTED** [review]
we need to lock `protobuf<=3.20.1` under `scripts/ci/config.yml#transformers`

**Review by aarnphm — APPROVED** [review]
Run test locally, now it is supported 😄

**Pull request merged**